tokio/src/runtime/basic_scheduler.rs

```diff
@@ -376,18 +376,23 @@ impl Context {
 
 impl Spawner {
     /// Spawns a future onto the basic scheduler
-    pub(crate) fn spawn<F>(&self, future: F, id: super::task::Id) -> JoinHandle<F::Output>
+    pub(crate) fn spawn<F>(
+        &self,
+        future: F,
+        id: super::task::Id,
+    ) -> (JoinHandle<F::Output>, Result<(), SpawnError>)
```

Comment:
Yeah in a number of places we currently return a `JoinHandle` even if we failed to schedule it because the runtime is shutting down and we can't break this API. It's kind of annoying but at least this is internal code which can be refactored if needed.

Returning the handle as part of the error is an interesting idea I hadn't considered! Though we would either need another wrapper since we wouldn't want the public type to have access to a `JoinHandle`, or hide it in some internal field. Let me play around with it and see if there is a nicer way to represent this!

Reply:
I added an internal wrapper which carries the handle. It made some places better, others worse, but net-net I kind of liked it. Let me know what you think! We can always revert it if needed